### ISSUE TITLE
Torch nondeterministic segfault

### ISSUE THREAD

**Issue by lukemetz** [dmlc/mxnet#1413]
Found another segfault with the `ThreadedEnginePerDevice` Executor. It does not happen on the `NaiveEngine`. It also appears to only happen when working with torch criterion. It appears to only occur when adding two criterion (MSE cost plus softmax for an autoencoder classifier model) after around 400 iterations of successful running.

Could there be a race in the order in which these two criterion are run  / issues if they are run at the same time?

My classification head looks something like:
```python
    flatten = some_big_model()
    fc1 = mx.symbol.FullyConnected(data=flatten, num_hidden=num_classes, name='fc1')

    label = mx.sym.Variable(name="softmax_label")

    logsoftmax = mx.symbol.TorchModule(data_0=fc1, lua_string='nn.LogSoftMax()', num_data=1, num_params=0, num_outputs=1, name='logsoftmax')
    label = mx.symbol.Variable('softmax_label') + 1
    softmax = mx.symbol.TorchCriterion(data=logsoftmax, label=label, lua_string='nn.ClassNLLCriterion()', name='softmax')
    recon = some_big_reconstruction_deconv(in5b)

    mse = mx.symbol.TorchCriterion(data=recon, label=data/150., lua_string='nn.MSECriterion()', name="mse", label_shape=(3, 64, 64))

    cost = softmax + mse*10

    D = lambda x: mx.symbol.BlockGrad(x)
    last = D(logsoftmax)

    final_symbol = mx.symbol.Group([cost])

```
Backtrace from GDB
```
Program received signal SIGSEGV, Segmentation fault.
[Switching to Thread 0x7ffe85ff3700 (LWP 29323)]
musable (mem=0x7ffe60891870) at malloc.c:4570
4570	malloc.c: No such file or directory.
(gdb) backtrace
Python Exception <class 'ImportError'> This package should not be accessible on Python 3. Either you are trying to run from the python-future src folder or your installation of python-future is corrupted.: 
#0  musable (mem=0x7ffe60891870) at malloc.c:4570
#1  __malloc_usable_size (m=0x7ffe60891870) at malloc.c:4581
#2  0x00007fffc138e029 in THFree ()
   from /home/luke/torch/install/lib/libTH.so
#3  0x00007fffbf5b9f08 in THCudaTensor_free ()
   from /home/luke/torch/install/lib/libTHC.so
#4  0x00007fffa9e15ca0 in torch_CudaTensor_free ()
   from /home/luke/torch/install/lib/lua/5.1/libcutorch.so
#5  0x00007fffc1a381fa in lj_BC_FUNCC ()
   from /home/luke/torch/install/lib/libluajit.so
#6  0x00007fffc19fa96a in gc_onestep ()
   from /home/luke/torch/install/lib/libluajit.so
#7  0x00007fffc19faabc in lj_gc_step ()
   from /home/luke/torch/install/lib/libluajit.so
#8  0x00007fffc1a21663 in lua_newuserdata ()
   from /home/luke/torch/install/lib/libluajit.so
#9  0x00007fffc17b2f7c in luaT_pushudata ()
   from /home/luke/torch/install/lib/libluaT.so
#10 0x00007fffccedddd6 in mxnet::TorchState::Deserialize(THCharStorage*) ()
   from /home/luke/Repos/dmlc_mxnet/python/mxnet/../../lib/libmxnet.so
#11 0x00007fffcd0e7136 in mxnet::op::TorchCriterionOp<mshadow::gpu>::Forward (this=0x385207b0, ctx=..., in_data=..., req=..., 
    out_data=..., aux_args=...)
    at plugin/torch/./torch_criterion-inl.h:79
#12 0x00007fffcce9da22 in mxnet::GraphExecutor::GetOpExecEntry(unsigned int)::{lambda(mxnet::RunContext, mxnet::engine::CallbackOnComplete)#1}::operator()(mxnet::RunContext, mxnet::engine::CallbackOnComplete) const ()
   from /home/luke/Repos/dmlc_mxnet/python/mxnet/../../lib/libmxnet.so
#13 0x00007fffcccbca89 in mxnet::engine::ThreadedEngine::ExecuteOprBlock(mxnet::RunContext, mxnet::engine::OprBlock*) ()
   from /home/luke/Repos/dmlc_mxnet/python/mxnet/../../lib/libmxnet.so
#14 0x00007fffcccbcf20 in std::_Function_handler<void (), mxnet::engine::ThreadedEnginePerDevice::PushToExecute(mxnet::engine::OprBlock*, bool)::{lambda()#3}::operator()() const::{lambda()#1}>::_M_invoke(std::_Any_data const&) ()
   from /home/luke/Repos/dmlc_mxnet/python/mxnet/../../lib/libmxnet.so
#15 0x00007fffbf236a40 in ?? ()
   from /usr/lib/x86_64-linux-gnu/libstdc++.so.6
#16 0x00007ffff7bc4182 in start_thread (arg=0x7ffe85ff3700)
    at pthread_create.c:312
#17 0x00007ffff78f147d in clone ()
    at ../sysdeps/unix/sysv/linux/x86_64/clone.S:111
```

Possibly related to: #1242

**Comment by piiswrong** [comment]
@mavenlin Maybe we should consider make separate lua_state per operator?

**Comment by lukemetz** [comment]
I will give it a try.

**Comment by piiswrong** [comment]
Actually there might be another cause. It failed in malloc and I think
malloc is non reentrant. So that might be the problem.

**Comment by lukemetz** [comment]
Also, why would a shared Lua_State matter if they are never used across threads? In both an operator specific Lua_State and thread specific Lua_State there should be no races.

**Comment by piiswrong** [comment]
You are probably right. So may be its non reentant malloc problem. Or
something else.

**Comment by lukemetz** [comment]
@piiswrong:
I have been trying to do a bit of investigating. My best guess as to what is going on is a race in when the lua garbage collector that is only getting triggered because an operator needs the output of two criterion.

I did a small test of putting a collectgarbage("collect") call into sharedserialize.lua save and load functions just before the return. This does create a reproducible segfault.

I am not sure why this is happening but is a problem.

**Comment by piiswrong** [comment]
is the lua garbage collect specific to lua states or a global background
thread?

**Comment by piiswrong** [comment]
What about two torch module instead of torch criterion?

**Comment by lukemetz** [comment]
I think the lua_state are completely separate and thus would have a separate garbage collector. (http://stackoverflow.com/questions/17653530/is-it-possible-to-share-a-table-between-multiple-lua-state-in-a-single-threaded)

The crash does appear to happen for both criterion and modules but its hard to 100% confirm. My test case is as follows:

```python
import mxnet as mx


def crash_combo_criterion():
    data = mx.sym.Variable(name="data")
    label = mx.sym.Variable(name="softmax_label")

    fc1 = mx.symbol.FullyConnected(data=data, num_hidden=10, name='fc1')
    logsoftmax = mx.symbol.TorchModule(data_0=fc1, lua_string='nn.LogSoftMax()', num_data=1, num_params=0, num_outputs=1, name='logsoftmax')
    # Torch's label starts from 1
    softmax = mx.symbol.TorchCriterion(data=logsoftmax, label=label+1, lua_string='nn.ClassNLLCriterion()', name='softmax')
    mse = mx.symbol.TorchCriterion(data=data, label=data+1, lua_string='nn.MSECriterion()', name="mse", label_shape=(10,))
    cost = softmax + mse*10

    executor = cost.simple_bind(data=(12, 10), softmax_label=(12,), ctx=mx.gpu(3))

    for i in range(1000):
        executor.forward()
        executor.backward()
        executor.outputs[0].asnumpy()

def crash_modules():
    data = mx.sym.Variable(name="data")

    fc1 = mx.symbol.FullyConnected(data=data, num_hidden=10, name='fc1')
    logsoftmax = mx.symbol.TorchModule(data_0=fc1, lua_string='nn.LogSoftMax()', num_data=1, num_params=0, num_outputs=1, name='logsoftmax')
    logsoftmax2 = mx.symbol.TorchModule(data_0=fc1, lua_string='nn.LogSoftMax()', num_data=1, num_params=0, num_outputs=1, name='logsoftmax')

    cost = logsoftmax + logsoftmax2

    executor = cost.simple_bind(data=(12, 10), ctx=mx.gpu(3))
    out = mx.nd.zeros((12, 10), ctx=mx.gpu(3))

    for i in range(1000):
        executor.forward()
        executor.backward([out])
        executor.outputs[0].asnumpy()

crash_combo_criterion()
#crash_modules()
```

I was unable to get these to crash however without a modification to the lua threads library.
The modification is to add a `collectgarbage("collect")` call just before the return in `serialize.load` and `serialize.save`. It is my understanding is telling lua to collect garbage should never release used instances and should never cause a crash.

The error message is slightly different than before
```

*** Error in `/usr/bin/python': realloc(): invalid pointer: 0x000000002d4be2a0 ***

Program received signal SIGABRT, Aborted.
0x00007ffff782dcc9 in __GI_raise (sig=sig@entry=6)
    at ../nptl/sysdeps/unix/sysv/linux/raise.c:56
56	../nptl/sysdeps/unix/sysv/linux/raise.c: No such file or directory.
(gdb) backtrace
Python Exception <class 'ImportError'> This package should not be accessible on Python 3. Either you are trying to run from the python-future src folder or your installation of python-future is corrupted.: 
#0  0x00007ffff782dcc9 in __GI_raise (sig=sig@entry=6)
    at ../nptl/sysdeps/unix/sysv/linux/raise.c:56
#1  0x00007ffff78310d8 in __GI_abort () at abort.c:89
#2  0x00007ffff786a394 in __libc_message (do_abort=do_abort@entry=1, 
    fmt=fmt@entry=0x7ffff7978b28 "*** Error in `%s': %s: 0x%s ***\n")
    at ../sysdeps/posix/libc_fatal.c:175
#3  0x00007ffff78750f7 in malloc_printerr (action=<optimized out>, 
    str=0x7ffff7974d5d "realloc(): invalid pointer", ptr=<optimized out>)
    at malloc.c:4996
#4  0x00007ffff787a192 in __GI___libc_realloc (oldmem=0x2d4be2a0, bytes=16)
    at malloc.c:2986
#5  0x00007fffc070b080 in THRealloc () from /home/luke/torch/install/lib/libTH.so
#6  0x00007fffc07218d4 in THFloatTensor_rawResize.constprop.246 ()
   from /home/luke/torch/install/lib/libTH.so
#7  0x00007fffa946f9d4 in THNN_FloatLogSoftMax_updateOutput ()
   from /home/luke/torch/install/lib/lua/5.1/libTHNN.so
#8  0x00007fffc0db71b9 in lj_vm_ffi_call ()
   from /home/luke/torch/install/lib/libluajit.so
#9  0x00007fffc0d8a2ef in lj_ccall_func ()
   from /home/luke/torch/install/lib/libluajit.so
#10 0x00007fffc0d8c976 in lj_cf_ffi_meta___call ()
   from /home/luke/torch/install/lib/libluajit.so
#11 0x00007fffc0db51fa in lj_BC_FUNCC ()
   from /home/luke/torch/install/lib/libluajit.so
#12 0x00007fffc0da429d in lua_pcall ()
   from /home/luke/torch/install/lib/libluajit.so
#13 0x00007fffccf408e9 in mxnet::op::TorchModuleProp::InferShape(std::vector<mshadow::TShape, std::allocator<mshadow::TShape> >*, std::vector<mshadow::TShape, std::allocator<mshadow::TShape> >*, std::vector<mshadow::TShape, std::allocator<mshadow::TShape> >*) const () from /home/luke/Repos/dmlc_mxnet/python/mxnet/../../lib/libmxnet.so
#14 0x00007fffccead136 in mxnet::StaticGraph::InferNodeShapes(std::vector<unsigned int, std::allocator<unsigned int> > const&, std::vector<std::vector<mshadow::TShape, std::allocator<mshadow::TShape> >, std::allocator<std::vector<mshadow::TShape, std::allocator<mshadow::TShape> > > >*, std::vector<std::vector<mshadow::TShape, std::allocator<mshadow::TShape> >, std::allocator<std::vector<mshadow::TShape, std::allocator<mshadow::TShape> > > >*) const ()
   from /home/luke/Repos/dmlc_mxnet/python/mxnet/../../lib/libmxnet.so
#15 0x00007fffcce8bc78 in mxnet::GraphExecutor::InitDataEntryInfo(std::vector<mxnet::NDArray, std::allocator<mxnet::NDArray> > const&, std::vector<mxnet::NDArray, std::allocator<mxnet::NDArray> > const&, std::vector<mxnet::OpReqType, std::allocator<mxnet::OpReqType> > const&, std::vector<mxnet::NDArray, std::allocator<mxnet::NDArray> > const&) () from /home/luke/Repos/dmlc_mxnet/python/mxnet/../../lib/libmxnet.so
#16 0x00007fffccea7568 in mxnet::GraphExecutor::Init(mxnet::Symbol, mxnet::Context const&, std::map<std::string, mxnet::Context, std::less<std::string>, std::allocator<std::pair<std::string const, mxnet::Context> > > const&, std::vector<mxnet::NDArray, std::allocator<mxnet::NDArray> > const&, std::vector<mxnet::NDArray, std::allocator<mxnet::NDArray> > const&, std::vector<mxnet::OpReqType, std::allocator<mxnet::OpReqType> > const&, std::vector<mxnet::NDArray, std::allocator<mxnet::NDArray> > const&) ()

```

**Comment by piiswrong** [comment]
Here is a temporary work around: Since crash happens in serialization/deserialzation, we can remove the need for them by making a lua_state per operator, instead of per thread. 
The down side of this is every operator will have to load a copy of torch. But hopefully torch lib doesn't take up that much memory.

**Comment by lukemetz** [comment]
Just submitted a PR that does just this. I gave up trying to track down the bad references. I think its something to do with the output variable on the nn.Module but I am not sure.
